### DISCUSSION TITLE
vbe 2 hardware scrolling

### DISCUSSION THREAD

**Discussion by defdefred** [VendelinSlezak/BleskOS#32] [Ideas]
I wonder if you ever tried the vbe function 6 and 7 to make a longer framebuffer and achieve ultra fast scrolling in the internet browser?
Could be possible to add an invisible border to the rigth to simplify edge mouse drawing...
Could be possible to have hidding panel around the screen too :-)

**Comment by VendelinSlezak** [comment]
> I wonder if you ever tried the vbe function 6 and 7 to make a longer framebuffer and achieve ultra fast scrolling in the internet browser?

Well, these functions would not help with scrolling. Bigger problem is that actual approach to scrolling in Internet browser is the slowest one :) It is because it redraws whole area. It was easy to program and it did not consume any more memory, so I just leaved it and went to other things. But I surely will reprogram it to move already rendered part of screen and draw only new part what will be much faster.

>Could be possible to add an invisible border to the rigth to simplify edge mouse drawing...
Could be possible to have hidding panel around the screen too :-)

I am kind of lost what exactly do you mean. Can you please explain it more? What should such panel contain?

**Comment by defdefred** [comment]
http://www.phatcode.net/res/221/files/vbe20.pdf
Imagine you have a 1280x1024 screen and a graphic card with 128MB of ram.

Internet Browser/text editor case:
You can configure VBE to present a 1280x26214 fb, so 25 screen long capacity.
Then you can render the whole web page one time to the long fb and use hardware scrolling to move the physical screen without any redraw.

Graphic editor case:
You can configure VBE to present a (100+1280)x1024 fb to have an hidden panel displaying graphic tools and color palette. When the mouse is going outside the screen to the right the panel popup and when the tool is chosen the panel disappear and the working area came back full screen.

Spreadsheet case:
You can configure VBE to presente a 5792x5792 fb to manage a big speadsheet and use hardware scrolling to move the physical screen without any redraw.

Seems that the maximum Scan Line Length and maximum Number of Scan Lines could be read from the vbe bios:

Input: AX = 4F06h VBE Set/Get Logical Scan Line Length
BL = 00h Set Scan Line Length in Pixels
= 01h Get Scan Line Length
= 02h Set Scan Line Length in Bytes
= 03h Get Maximum Scan Line Length
CX = If BL=00h Desired Width in Pixels
If BL=02h Desired Width in Bytes
(Ignored for Get Functions
Output: AX = VBE Return Status
BX = Bytes Per Scan Line
CX = Actual Pixels Per Scan Line
(truncated to nearest complete pixel)
DX = Maximum Number of Scan Lines

And the function 7 could be used in 32bit protected mode:

Input: AX = 4F0Ah VBE 2.0 Protected Mode Interface
BL = 00h Return protected mode table
Output: AX = Status
ES = Real Mode Segment of Table
DI = Offset of Table
CX = Length of Table including protected mode code in bytes
(for copying purposes)
VBE Functions
0Ah Return VBE Protected Mode Interface
Page 32 VBE CORE FUNCTIONS VERSION 2.0
DOCUMENT REVISION 1.1
The format of the table is as follows:
ES:DI + 00h Word Offset in table of Protected mode code for
Function 5 for Set Window Call
ES:DI + 02h Word Offset in table of Protected mode code for
Function 7 for set Display Start

**Comment by VendelinSlezak** [comment]
Thank you for explanation. Maybe there would be some improvement by using VBE, but overall I think all those are much simpler to achieve by storing rendered image in basic RAM. Also function 7 seems to rerender whole screen so for scrolling there would be need to store borders and rerender them every time screen is moved what could be visible as blinking for user. It seems to me that this function is best for changing whole screen to something else.

**Comment by defdefred** [comment]
And the big issue is that vbe "hard scrolling" function have no GOP equivalent for uefi/bleskos64 :-).

Look likes everybody is using double buffer in standard ram to avoid slow read in graphics memory.

**Comment by VendelinSlezak** [comment]
Yes, these are another two valid points. BleskOS also use double buffering for graphic.
